src/Marten/Services/JsonNetSerializer.cs

```diff
@@ -95,20 +109,31 @@ public T FromJson<T>(Stream stream)
             return _serializer.Deserialize<T>(jsonReader);
         }
 
-        public T FromJson<T>(TextReader reader)
+        public ValueTask<T> FromJsonAsync<T>(Stream stream, CancellationToken cancellationToken = default)
+        {
+            return new(FromJson<T>(stream));
+        }
+
+        public ValueTask<T> FromJsonAsync<T>(DbDataReader reader, int index, CancellationToken cancellationToken = default)
+        {
+            return new(FromJson<T>(reader, index));
+        }
+
+        public object FromJson(Type type, Stream stream)
         {
-            using var jsonReader = new JsonTextReader(reader)
+            using var jsonReader = new JsonTextReader(stream.GetStreamReader())
             {
                 ArrayPool = _jsonArrayPool,
                 CloseInput = false
             };
 
-            return _serializer.Deserialize<T>(jsonReader);
+            return _serializer.Deserialize(jsonReader, type);
         }
 
-        public object FromJson(Type type, TextReader reader)
+        public object FromJson(Type type, DbDataReader reader, int index)
         {
-            using var jsonReader = new JsonTextReader(reader)
+            using var textReader = reader.GetTextReader(index);
```

Comment:
For later. If we took in the native Npgsql DbDataReader type instead, you get access to async versions of `GetTextReader()` and `GetStream()`